modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -4253,15 +4258,40 @@ def iloc_mut(partition, row_internal_indices, col_internal_indices, item):
                 partition.iloc[row_internal_indices, col_internal_indices] = item.copy()
             return partition
 
+        if not is_scalar(item):
+            broadcasted_item, broadcasted_dtypes = broadcast_item(
+                self,
+                row_numeric_index,
+                col_numeric_index,
+                item,
+                need_columns_reindex=need_columns_reindex,
+            )
+        else:
+            broadcasted_item, broadcasted_dtypes = item, pandas.Series(
+                [np.array(item).dtype] * len(col_numeric_index)
+            )
+
+        new_dtypes = None
+        if (
+            # compute dtypes only if assigning entire columns
+            isinstance(row_numeric_index, slice)
+            and row_numeric_index == slice(None)
+            and self._modin_frame.has_materialized_dtypes
+            and broadcasted_dtypes is not None
```

Comment:
I would suggest move this check to be the very first in this condition, as it's the cheapest to compute (super-tiny optimization which is probably useless here but could be a good habit of writing new code that way - from cheapest to heavier, to determine the most probable result with as small cost as possible)

Reply:
agree, done